[dtensor] add async_op option to redistribute and some refactor

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121477

async output option was only available in `full_tensor()` call, but I think it's
generally good to make this option available in the `redistribute` call directly
so that user can control it

This PR adds async_op option to redistribute call, to allow user control
whether to perform tensor redistribution asynchronously or not.

By default we set this to False, this is to follow the semantics of the c10d
collectives.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

wanchaol committed:
[dtensor] add async_op option to redistribute and some refactor

This PR adds async_op option to redistribute call, to allow user control
whether to perform tensor redistribution asynchronously or not.

By default we set this to False, this is to follow the semantics of the c10d
collectives.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add async_op option to redistribute and some refactor"


async output option was only available in `full_tensor()` call, but I think it's
generally good to make this option available in the `redistribute` call directly
so that user can control it

This PR adds async_op option to redistribute call, to allow user control
whether to perform tensor redistribution asynchronously or not.

By default we set this to False, this is to follow the semantics of the c10d
collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add async_op option to redistribute and some refactor"


async output option was only available in `full_tensor()` call, but I think it's
generally good to make this option available in the `redistribute` call directly
so that user can control it

This PR adds async_op option to redistribute call, to allow user control
whether to perform tensor redistribution asynchronously or not.

By default we set this to False, this is to follow the semantics of the c10d
collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wz337 commented on test/distributed/tensor/parallel/test_tp_examples.py:
```diff
@@ -94,11 +94,16 @@ def _test_mlp_training_e2e(self, is_seq_parallel=False, recompute_activation=Fal
+            output_tp.sum().backward()
+
+        self.assertEqual(output, output_tp)
+        self.assertEqual(comm_mode.get_comm_counts()[c10d_functional.all_reduce], 1)
```
Seems the test is complaining about this check for `test_mlp_training`.

wanchaol replied:
yeah adjusted

wz337 commented:
Seems this test also needs to be adjusted for the change. 
https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_dtensor.py#L433

wanchaol committed:
Update on "[dtensor] add async_op option to redistribute and some refactor"


async output option was only available in `full_tensor()` call, but I think it's
generally good to make this option available in the `redistribute` call directly
so that user can control it

This PR adds async_op option to redistribute call, to allow user control
whether to perform tensor redistribution asynchronously or not.

By default we set this to False, this is to follow the semantics of the c10d
collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wz337 reviewed (approved):
LGTM!

wanchaol commented:
@pytorchbot merge
